homeassistant/components/stream/core.py

```diff
@@ -356,3 +359,46 @@ async def handle(
     ) -> web.StreamResponse:
         """Handle the stream request."""
         raise NotImplementedError()
+
+
+class KeyFrame:
+    """Hold the last keyframe.
+
+    The keyframe can be either a Packet or a jpeg of bytes. This is so that a
+    new keyframe assignment can remain atomic to avoid threading issues.
+    """
+
+    def __init__(self) -> None:
+        """Initialize."""
+
+        # Keep import here so that we can import stream integration without installing reqs
+        # pylint: disable=import-outside-toplevel
+        from homeassistant.components.camera.img_util import TurboJPEGSingleton
+
+        self.keyframe: Packet | bytes | None = None
+        self.turbojpeg = TurboJPEGSingleton.instance()
+        self.get_bytes = self._get_bytes
+        if not self.turbojpeg:
+            self.get_bytes = lambda: None
+
+    def _get_bytes(self) -> bytes | None:
+        """Get the keyframe as bytes."""
+
+        # Keep import here so that we can import stream integration without installing reqs
+        # pylint: disable=import-outside-toplevel
+        from av import Packet
+
+        last_packet_or_image = self.keyframe
+        if isinstance(last_packet_or_image, Packet):
+            # decode packet (try up to 3 times)
+            for _i in range(3):
+                # pylint: disable=maybe-no-member
+                if frames := last_packet_or_image.decode():
```

Comment:
Have details of the crash? I assume its in turbojpeg, but i was curious about this step.  I was thinking maybe this step is the problem since its produced by the worker thread but run in the main loop.  However reading https://pyav.org/docs/develop/cookbook/basics.html#threading it didn't necessarily give me the impression this decode step could be decoupled, or that its cheap enough we can do it all the time. That is, it seems both heavyweight and tied to worker state.

How often are there keyframes?

(I was thinking through an alternative where the Stream asks the worker for a frame and it produces it on demand as it iterates, which is a little more complex from an orchestration perspective, but maybe less complex from a thread safety perspective if its just passing off the final bytes between threads)

Reply:
No, there wasn't much in my docker logs - I think the process exited with a `256` error.
After looking at a discussion about threading in libjpeg-turbo, I don't think it's from that side. It might be from the python wrapper, but there's a good chance it's from the pyav side.
I haven't been able to repro. It happened when I had 10+ cameras on the same page pulling images every 10 seconds. I don't have a separate dev instance so I took off those settings and only have it pulling on one camera right now, and I haven't seen any problems.

Keyframes depend on the input stream. Generally they're every 1 second or every few seconds, but for some of the modified codecs like H264+ H265+ they use really long intervals like 20 seconds to save on bandwidth. I use those myself, and I don't mind the lag. With the regular keyframe intervals this is not a concern, but for long keyframe intervals an alternative implementation is to wait for the _next_ keyframe instead of the last one, resulting in a longer time to the first request but a shorter "lag" between the displayed picture and real time.
We could try to move it to the worker, and that might solve a threading issue if we have one, but we'd have to add the overhead of the worker to core events/signaling. Currently the only inter thread interaction is when the packet is set, but I think that is atomic.
One more feature we can add is options for quality/scaling which can help to speed up the encode/implement the width/height parameters used in `Camera.async_camera_image`